engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/bigInteger/PowNode.java

```diff
@@ -0,0 +1,53 @@
+package org.enso.interpreter.node.expression.builtin.number.bigInteger;
+
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.node.expression.builtin.number.utils.BigIntegerOps;
+import org.enso.interpreter.node.expression.builtin.number.utils.ToEnsoNumberNode;
+import org.enso.interpreter.runtime.error.TypeError;
+import org.enso.interpreter.runtime.number.EnsoBigInteger;
+
+@BuiltinMethod(type = "Big_Integer", name = "^", description = "Big integer exponentiation.")
+public abstract class PowNode extends Node {
+  private @Child ToEnsoNumberNode toEnsoNumberNode = ToEnsoNumberNode.build();
+
+  public abstract Object execute(EnsoBigInteger _this, Object that);
+
+  public static PowNode build() {
+    return PowNodeGen.create();
+  }
+
+  @Specialization
+  Object doLong(EnsoBigInteger _this, long that) {
+    if (that == 0) {
+      return 1L;
+    } else if (that > 0) {
+      return toEnsoNumberNode.execute(BigIntegerOps.pow(_this.getValue(), that));
+    } else {
+      return Math.pow(BigIntegerOps.toDouble(_this.getValue()), that);
+    }
+  }
+
+  @Specialization
+  Object doBigInteger(EnsoBigInteger _this, EnsoBigInteger that) {
+    if (that.getValue().signum() > 0) {
+      return Double.POSITIVE_INFINITY;
```

Comment:
I think that this branch should also rely on the sign of `_this`.

`x^M < 0 if x < 0 and M % 2 == 1`

Reply:
Also while it seems reasonable that for such big values there is no sense in computing the exact result as (I haven't done the math but feels like) it may not even fit in memory, but I think it would be good if the intricacies of Enso's number system were documented - it is non obvious when Integers can become Decimals etc. - it is already documented for `/` so I think a similar paragraph about `^` would also be useful.